common/array_stack.h

```diff
@@ -0,0 +1,74 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_COMMON_ARRAY_STACK_H_
+#define CARBON_COMMON_ARRAY_STACK_H_
+
+#include "common/check.h"
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/SmallVector.h"
+
+namespace Carbon {
+
+// Provides a stack of arrays. Only the array at the top of the stack can have
+// elements added.
+//
+// Example usage:
+//   // Push to start.
+//   PushArray();
+//   // Add values.
+//   PushValue(3);
+//   // Look at values.
+//   PeekArray();
+//   // Pop when done.
+//   PopArray();
+//
+// By using a single vector for elements, the intent is that as arrays are
+// pushed and popped, the same storage will be reused. This should yield
+// efficiencies for heap allocations. For example, in the toolchain we
+// frequently have an array per scope, and only add to the current scope's
+// array; this allows better reuse when entering and leaving scopes.
+template <typename ValueT>
+class ArrayStack {
+ public:
+  // Pushes a new array onto the stack.
+  auto PushArray() -> void { array_offsets_.push_back(elements_.size()); }
+
+  // Pops the top array from the stack.
+  auto PopArray() -> void {
+    auto region = array_offsets_.pop_back_val();
+    elements_.truncate(region);
+  }
+
+  // Returns the top array from the stack.
+  auto PeekArray() const -> llvm::ArrayRef<ValueT> {
+    CARBON_CHECK(!array_offsets_.empty());
+    return llvm::ArrayRef(elements_).slice(array_offsets_.back());
+  }
+
+  // Returns the full set of values on the stack, regardless of whether any
+  // arrays are pushed.
+  auto PeekAllValues() const -> llvm::ArrayRef<ValueT> { return elements_; }
+
+  // Adds a value to the top array on the stack.
+  auto PushValue(ValueT value) -> void {
```

Comment:
Calling this `Push...` makes the API a bit harder for me tu understand because there isn't a pop and it doesn't interact with the stack-based semantic really...

WDYT about `AddToTop` or `AppendToTop`?

Also open to other words besides `Add` or `Append` here, just trying to find a way to keep it separate from the push/pop/peek done to the arrays.

Also open to other words besides `Top` -- maybe `AddToArray`?

Anyways, curious what you think might work well naming wise.

Reply:
Renamed to `AppendToTop`

`AddToArray` doesn't work for me well because it makes me wonder _which_ array. The implicit aspect of "top" is not clear to me once you start talking about `Array`, unlike `Peek`. I'm not really wanting to say `AddToTopArray`.